File: Greedy/ActivitySelection.java
import java.util.*;
import java.io.*;

/*
Activity Selection
Given N activities with their start and finish times. Select the maximum number of activities that can be performed by a single person, assuming that a person can only work on a single activity at a time.

Note : The start time and end time of two activities may coincide.

Input:
The first line contains T denoting the number of testcases. Then follows description of testcases. First line is N number of activities then second line contains N numbers which are starting time of activies.Third line contains N finishing time of activities.

Output:
For each test case, output a single number denoting maximum activites which can be performed in new line.

Constraints:
1<=T<=50
1<=N<=1000
1<=A[i]<=100

Example:
Input:
2
6
1 3 2 5 8 5
2 4 6 7 9 9
4
1 3 2 5
2 4 3 6

Output:
4
4
*/

class Activity
{
    int start, end, pos;
    
    Activity(int s, int e, int p)
    {
        start = s;
        end = e;
        pos = p;
    }
}

class CompareByEnd implements Comparator<Activity>
{
    public int compare(Activity a, Activity b)
    {
        return a.end - b.end;   
    }
}

class ActivitySelection
{
    static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
    public static void main (String[] args) throws Exception
	{
	    int T = Integer.parseInt(bf.readLine().trim());
	    for(int a=0; a<T; a++)
	    {
	        int N = Integer.parseInt(bf.readLine().trim());
	        String str1[] = bf.readLine().split(" ");
	        String str2[] = bf.readLine().split(" ");
	        
	        ArrayList<Activity> arr = new ArrayList<>();
	        for(int i=0; i<N; i++)
	        {
	            int start = Integer.parseInt(str1[i]);
	            int end = Integer.parseInt(str2[i]);
	            
	            Activity at = new Activity(start, end, i);
	            arr.add(at);
            }
            Collections.sort(arr, new CompareByEnd());
            
            StringBuffer sb = new StringBuffer();
            
            int count = 0;
            int lastDoneEndTime = -1;
            for(int i=0; i<N; i++)
            {
                if(lastDoneEndTime == -1)//startting 1st activity
                {
                    lastDoneEndTime = arr.get(i).end;
                    count++;
                    
                    sb.append(arr.get(i).pos+1 + " ");
                }
                else
                {
                    if(arr.get(i).start >= lastDoneEndTime)
                    {
                        lastDoneEndTime = arr.get(i).end;
                        count++;
                        
                        sb.append(arr.get(i).pos+1 + " ");
                    }
                }
            }
            //System.out.println(sb);
            System.out.println(count);
	    }
	}
}

File: JS/greedy/activitySelection.js
/*
https://practice.geeksforgeeks.org/problems/activity-selection-1587115620/1

Activity Selection
Join the most popular course on DSA. Master Skills & Become Employable by enrolling today! 
Given N activities with their start and finish day given in array start[ ] and end[ ]. Select the maximum number of activities that can be performed by a single person, assuming that a person can only work on a single activity at a given day.
Note : Duration of the activity includes both starting and ending day.


Example 1:

Input:
N = 2
start[] = {2, 1}
end[] = {2, 2}
Output: 
1
Explanation:
A person can perform only one of the
given activities.
Example 2:

Input:
N = 4
start[] = {1, 3, 2, 5}
end[] = {2, 4, 3, 6}
Output: 
3
Explanation:
A person can perform activities 1, 2
and 4.

Your Task :
You don't need to read input or print anything. Your task is to complete the function activityselection() which takes array start[ ], array end[ ] and integer N as input parameters and returns the maximum number of activities that can be done.


Expected Time Complexity : O(N * Log(N))
Expected Auxilliary Space : O(N)
*/


class Solution {
    //Function to find the maximum number of activities that can
    //be performed by a single person.
    activitySelection(startArr, endArr, n) {
        // making a single array for startArr & endArr and sorting it by end
        // because we will firstly do activities that are ending earlier so that we can do maximum activities
        const arr = startArr.map((start, idx) => ({ start, end: endArr[idx] }));
        arr.sort((a, b) => a.end - b.end);

        let prevEnd = arr[0].end; // we will start with the activity with least end time
        let c = 1;
        for (let i = 1; i < n; i++) {
            const { start: thisstart, end: thisEnd } = arr[i];

            if (thisstart > prevEnd) { // if this activity starts strictly after end time of the last/prev activity // only then man can do the activity
                c++;
                prevEnd = thisEnd;
            }
        }

        return c;
    }
}